File: Data_Structure/Python/UndirectedGraph.py
from collections import deque

class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.adj_list = {i: [] for i in range(vertices)}

    # Add undirected edge
    def add_edge(self, src, dest):
        if dest not in self.adj_list[src]:
            self.adj_list[src].append(dest)
        if src not in self.adj_list[dest]:
            self.adj_list[dest].append(src)

    # Remove undirected edge
    def remove_edge(self, src, dest):
        if dest in self.adj_list[src]:
            self.adj_list[src].remove(dest)
        if src in self.adj_list[dest]:
            self.adj_list[dest].remove(src)

    # Display the graph
    def display_graph(self):
        print("\n--- Graph Adjacency List ---")
        for vertex, neighbors in self.adj_list.items():
            print(f"{vertex}: {' -> '.join(map(str, neighbors))}")

    # DFS traversal
    def dfs(self, start):
        visited = [False] * self.V
        print(f"\nDFS Traversal starting from vertex {start}: ", end="")
        self._dfs_util(start, visited)
        print()

    def _dfs_util(self, vertex, visited):
        visited[vertex] = True
        print(vertex, end=" ")

        for neighbor in self.adj_list[vertex]:
            if not visited[neighbor]:
                self._dfs_util(neighbor, visited)

    # BFS traversal
    def bfs(self, start):
        visited = [False] * self.V
        queue = deque()

        visited[start] = True
        queue.append(start)

        print(f"\nBFS Traversal starting from vertex {start}: ", end="")

        while queue:
            vertex = queue.popleft()
            print(vertex, end=" ")

            for neighbor in self.adj_list[vertex]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    queue.append(neighbor)
        print()


# Main program
if __name__ == "__main__":
    V = int(input("Enter number of vertices: "))
    graph = Graph(V)

    while True:
        print("\n--- Undirected Graph Operations ---")
        print("1. Add Edge")
        print("2. Remove Edge")
        print("3. Display Graph")
        print("4. DFS Traversal")
        print("5. BFS Traversal")
        print("6. Exit")

        choice = int(input("Enter your choice: "))

        if choice == 1:
            src = int(input("Enter source: "))
            dest = int(input("Enter destination: "))
            graph.add_edge(src, dest)

        elif choice == 2:
            src = int(input("Enter source: "))
            dest = int(input("Enter destination: "))
            graph.remove_edge(src, dest)

        elif choice == 3:
            graph.display_graph()

        elif choice == 4:
            start = int(input("Enter starting vertex for DFS: "))
            graph.dfs(start)

        elif choice == 5:
            start = int(input("Enter starting vertex for BFS: "))
            graph.bfs(start)

        elif choice == 6:
            print("Exiting...")
            break

        else:
            print("Invalid choice! Please try again.")

File: Data_Structure/C/UndirectedGraph.c
#include <stdio.h>
#include <stdlib.h>

// Structure for adjacency list node
struct Node {
    int vertex;
    struct Node* next;
};

// Structure for adjacency list
struct AdjList {
    struct Node* head;
};

// Structure for graph
struct Graph {
    int numVertices;
    struct AdjList* array;
};

// Function to create a new adjacency list node
struct Node* createNode(int v) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->vertex = v;
    newNode->next = NULL;
    return newNode;
}

// Function to create a graph
struct Graph* createGraph(int vertices) {
    struct Graph* graph = (struct Graph*)malloc(sizeof(struct Graph));
    graph->numVertices = vertices;

    graph->array = (struct AdjList*)malloc(vertices * sizeof(struct AdjList));

    for (int i = 0; i < vertices; i++)
        graph->array[i].head = NULL;

    return graph;
}

// Function to add an undirected edge
void addEdge(struct Graph* graph, int src, int dest) {
    // Add edge src -> dest
    struct Node* newNode = createNode(dest);
    newNode->next = graph->array[src].head;
    graph->array[src].head = newNode;

    // Add edge dest -> src (because undirected)
    newNode = createNode(src);
    newNode->next = graph->array[dest].head;
    graph->array[dest].head = newNode;
}

// Function to remove an edge
void removeEdge(struct Graph* graph, int src, int dest) {
    struct Node *temp = graph->array[src].head, *prev = NULL;
    while (temp != NULL && temp->vertex != dest) {
        prev = temp;
        temp = temp->next;
    }
    if (temp != NULL) {
        if (prev == NULL)
            graph->array[src].head = temp->next;
        else
            prev->next = temp->next;
        free(temp);
    }

    // Remove dest -> src
    temp = graph->array[dest].head;
    prev = NULL;
    while (temp != NULL && temp->vertex != src) {
        prev = temp;
        temp = temp->next;
    }
    if (temp != NULL) {
        if (prev == NULL)
            graph->array[dest].head = temp->next;
        else
            prev->next = temp->next;
        free(temp);
    }
}

// Display the graph
void displayGraph(struct Graph* graph) {
    for (int v = 0; v < graph->numVertices; v++) {
        struct Node* temp = graph->array[v].head;
        printf("\nAdjacency list of vertex %d\n head", v);
        while (temp) {
            printf(" -> %d", temp->vertex);
            temp = temp->next;
        }
        printf("\n");
    }
}

// DFS utility
void DFSUtil(struct Graph* graph, int vertex, int visited[]) {
    visited[vertex] = 1;
    printf("%d ", vertex);

    struct Node* temp = graph->array[vertex].head;
    while (temp) {
        int adjVertex = temp->vertex;
        if (!visited[adjVertex])
            DFSUtil(graph, adjVertex, visited);
        temp = temp->next;
    }
}

// DFS traversal
void DFS(struct Graph* graph, int startVertex) {
    int* visited = (int*)malloc(graph->numVertices * sizeof(int));
    for (int i = 0; i < graph->numVertices; i++)
        visited[i] = 0;

    printf("\nDFS Traversal starting from vertex %d: ", startVertex);
    DFSUtil(graph, startVertex, visited);
    printf("\n");
    free(visited);
}

// BFS traversal
void BFS(struct Graph* graph, int startVertex) {
    int* visited = (int*)malloc(graph->numVertices * sizeof(int));
    for (int i = 0; i < graph->numVertices; i++)
        visited[i] = 0;

    int queue[100], front = 0, rear = 0;

    visited[startVertex] = 1;
    queue[rear++] = startVertex;

    printf("\nBFS Traversal starting from vertex %d: ", startVertex);

    while (front < rear) {
        int currentVertex = queue[front++];
        printf("%d ", currentVertex);

        struct Node* temp = graph->array[currentVertex].head;
        while (temp) {
            int adjVertex = temp->vertex;
            if (!visited[adjVertex]) {
                visited[adjVertex] = 1;
                queue[rear++] = adjVertex;
            }
            temp = temp->next;
        }
    }
    printf("\n");
    free(visited);
}

// Main function
int main() {
    int V, choice, src, dest, start;

    printf("Enter number of vertices: ");
    scanf("%d", &V);

    struct Graph* graph = createGraph(V);

    while (1) {
        printf("\n--- Undirected Graph Operations ---\n");
        printf("1. Add Edge\n2. Remove Edge\n3. Display Graph\n4. DFS\n5. BFS\n6. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
        case 1:
            printf("Enter source and destination: ");
            scanf("%d %d", &src, &dest);
            addEdge(graph, src, dest);
            break;
        case 2:
            printf("Enter source and destination: ");
            scanf("%d %d", &src, &dest);
            removeEdge(graph, src, dest);
            break;
        case 3:
            displayGraph(graph);
            break;
        case 4:
            printf("Enter starting vertex for DFS: ");
            scanf("%d", &start);
            DFS(graph, start);
            break;
        case 5:
            printf("Enter starting vertex for BFS: ");
            scanf("%d", &start);
            BFS(graph, start);
            break;
        case 6:
            exit(0);
        default:
            printf("Invalid choice\n");
        }
    }

    return 0;
}